src/java.base/share/classes/javax/crypto/JceSecurityManager.java

```diff
@@ -232,13 +238,17 @@ private CryptoPermission getDefaultPermission(String alg) {
     // objects being constructed by untrusted code (See bug 4341369 &
     // 4334690 for more info).
     boolean isCallerTrusted(Provider provider) {
-        // Get the caller and its codebase.
-        Class<?>[] context = getClassContext();
-        if (context.length >= 3) {
+        PrivilegedAction<StackWalker> pa =
+                () -> StackWalker.getInstance(Option.RETAIN_CLASS_REFERENCE);
+        @SuppressWarnings("removal")
+        Optional<StackFrame> stackFrame = AccessController.doPrivileged(pa)
+                .walk((s) -> s.skip(2).findFirst());
```

Comment:
You can use the same `StackWalker` instance in multiple places. 

`StackWalker::getCallerClass` is the API to get the caller class.   You want to get the caller of the subclass of `Cipher` in this case.    So `Cipher` constructor will call `walker.getCallerClass()` and then pass it to `isCallerTrusted` which will take an additional caller class parameter for validation.